hr_timesheet_holiday/models/hr_holidays.py

```diff
@@ -0,0 +1,161 @@
+
+    analytic_line_ids = fields.One2many(
+        "account.analytic.line", "leave_id", "Analytic Lines"
+    )
+
+    @api.multi
+    def add_timesheet_line(self, description, date, hours, account):
+        """Add a timesheet line for this leave"""
+        self.ensure_one()
+        # User exists because already check during the holidays_validate
+        user = self.employee_id.user_id
+        self.sudo().with_context(force_write=True).write(
+            {
+                "analytic_line_ids": [
+                    (
+                        0,
+                        False,
+                        {
+                            "name": description,
+                            "date": date,
+                            "unit_amount": hours,
+                            "company_id": self.employee_id.company_id.id,
+                            "account_id": account.id,
+                            "user_id": user.id,
+                            "is_timesheet": True,
+                        },
+                    )
+                ]
+            }
+        )
+
+    @api.model
+    def _get_company_hours_per_day(self, company):
+        hours_per_day = company.timesheet_hours_per_day
+        if not hours_per_day:
+            raise UserError(
+                _("No hours per day defined for Company '%s'")
+                % (company.name,)
+            )
+        return hours_per_day
+
+    @api.model
+    def _get_contract_hours_per_day(self, employee, date):
+        self.ensure_one()
+        start_dt = date
+        end_dt = False
+        from_dt = self._compute_datetime(self.date_from)
+        to_dt = self._compute_datetime(self.date_to)
+        if date.date() == from_dt.date():
+            start_dt = from_dt
+        elif date.date() == to_dt.date():
+            end_dt = to_dt
+        hours_per_day = 0.0
+        contracts = (
+            self.env["hr.contract"]
+            .sudo()
+            .search([("employee_id.id", "=", employee.id)])
+        )
+        for contract in contracts:
+            for calendar in contract.working_hours:
+                if end_dt:
+                    working_hours = calendar.get_working_hours_of_date(start_dt=start_dt, end_dt=end_dt)
+                else:
+                    working_hours = calendar.get_working_hours_of_date(start_dt=start_dt)
+                for wh in working_hours:
+                    hours_per_day += wh
+        return hours_per_day
+
+    @api.model
+    def _compute_datetime(self, date):
+        dt = False
+        if date:
+            this_year = datetime.date.today().year
+            reference_date = fields.Datetime.context_timestamp(
+                self.env.user, datetime.datetime(this_year, 1, 1, 12)
+            )
+            dt = fields.Datetime.from_string(date)
+            tz_dt = fields.Datetime.context_timestamp(self.env.user, dt)
+            dt = dt + tz_dt.tzinfo._utcoffset
+            dt = dt - reference_date.tzinfo._utcoffset
```

Comment:
I don't understand this, sorry. :/

Reply:
i removed this method since `fields.Datetime.from_string(self.date_from)` does the job to transform a string date into a datetime object